Add a prominent summary section (#48)
WICG/app-history: README.md

This new `window.appHistory` API [layers](#integration-with-the-existing-history-api-and-spec) on top of the existing API and specification infrastructure, with well-defined interaction points. The main differences are that it is scoped to the current origin and frame, and it is designed to be pleasant to use instead of being a historical accident with many sharp edges.

## Summary

There's a lot going on here, and this explainer tries to really explain it! Here's the short version.

> The existing [history API](https://developer.mozilla.org/en-US/docs/Web/API/History) is bad, especially for single-page applications. This leads to user-facing pain in the form of lost state and broken back buttons, and developer-facing pain in the form of maintaining lots of hacky code to achieve the desired user experience.

Reviewer: ```suggestion
The existing [history API](https://developer.mozilla.org/en-US/docs/Web/API/History) is hard to deal with in practice, especially for single-page applications. This leads to user-facing pain in the form of lost state and broken back buttons, and developer-facing pain in the form of maintaining lots of hacky code to achieve the desired user experience.
```

A key reason for this new API, as opposed to extending the `window.history` API, is because we need to avoid `window.history`'s original sin: its over-broad scope. In particular, if an iframe navigates, this contributes to the user's view of session history, e.g. what happens when they press the back button. But for application code, this is not helpful: iframe navigations are generally independent of the application, so the fact that they can mess with session history makes `history.back()` and `history.state` unreliable. Similarly, because clicking back can cause the user to move across origins, `window.history` includes cross-origin information, which means we can't add better introspection facilities (such as the [previously-proposed `history.index`](https://github.com/whatwg/html/issues/2710)) without leaking cross-origin information. App history separates the user's view of history from the web application's, by giving a view onto the same-origin, same-frame, contiguous history entries only. This opens the door to allow in-depth introspection and manipulation of the history list.

> 

Reviewer 2: This one is just too wordy. It's detailed in a summary!

I think something like this instead:

- `appHistory` provides a same-origin, same-frame, contiguous history view of the browser history. History entries added by iframes are not visible through `appHistory`, and no events will be sent for those navigations. This fixes longstanding issues with iframes polluting the history stack and opens the door for in-depth introspection and manipulations of application owned history.

> Additionally, when single-page navigations are implemented by intercepting the `navigate` event, they become first-class citizens. Browser features like the loading spinner, or accessibility technology, can key off of this standardized entry point. And other APIs like performance timing APIs can measure the duration of such navigations.

Reviewer 2: ```suggestion
`appHistory` will enable browsers to implement features like the loading spinner or accessibility technology for in-app navigations. And other APIs like performance timing APIs can measure the duration of such navigations more accurately.
```

- Some of the history APIs are clunky and hard to understand. We solve this by providing a new interface that is easy for developers to use and understand.

## Sample code

An application or framework's centralized router can use the `navigate` event to implement single-page app routing:

```js
appHistory.addEventListener("navigate", e => {
  if (!e.sameOrigin || e.hashChange) {
    return;
  }

  if (routesTable.has(e.destination.url)) {
    const routeHandler = routesTable.get(e.destination.url);
    e.respondWith(routeHandler());
  }
});
```

A page-supplied "back" button can actually take you back, even after reload, by inspecting the previous history entries:

```js
backButtonEl.addEventListener("click", () => {
  if (appHistory.entries[appHistory.entries.current.index - 1]?.url === "/product-listing") {
    // Go back to that entry
    appHistory.back();
  } else {
    // Go "forward" to that entry, e.g. if the user arrived here by typing the URL directly.

>     appHistory.push("/product-listing");

Reviewer 2: More likely this would be a replace?
